torch/csrc/autograd/record_function.cpp

```diff
@@ -189,14 +193,60 @@ void RecordFunction::processCallbacks() {
 }
 
 RecordFunction::~RecordFunction() {
+  end();
+}
+
+void RecordFunction::end() {
   if (initialized_) {
     for (size_t idx = 0; idx < manager().end_callbacks.size(); ++idx) {
       if (!manager().is_callback_sampled[idx] || run_sampled_) {
         manager().end_callbacks[idx](*this);
       }
     }
+
+    AT_ASSERT(thread_local_func_ == this, name_, ": must be top of stack");
     thread_local_func_ = parent_;
+    initialized_ = false;
   }
 }
 
-}}}
+at::Tensor record_function_enter(const std::string& name) {
+  auto rec = at::guts::make_unique<RecordFunction>();
+  // Only add new scope if profiling is enabled.
+  if (thread_local_func_) {
+    AT_ASSERT(
+        thread_local_func_->name() ==
+        StringView("profiler::record_function_enter"));
+    // RecordFunction requires parent_ to be alive for it's entire lifetime.
+    // Since the currently active RecordFunction will only live for the lifetime
+    // of this op we need to end it early so the new RecordFunction we create is
+    // a direct child of the parent RecordFunction.
+    thread_local_func_->end();
+
+    rec->before(name);
+  }
+  return at::cpp_custom_type_hack::create(std::move(rec), at::TensorOptions());
+}
+
+void record_function_exit(const at::Tensor& handle) {
+  // We don't actually need to do anything with handle just need to persist the
+  // lifetime until now.
+  auto& rec = at::cpp_custom_type_hack::cast<RecordFunction>(handle);
+  if (thread_local_func_) {
+    AT_ASSERT(thread_local_func_->parent() == &rec, "rec must be parent");
+    AT_ASSERT(
+        thread_local_func_->name() ==
+        StringView("profiler::record_function_exit"));
+    thread_local_func_->end();
```

Comment:
nit: I'm wondering if it's better to have the scope event ended earlier than record_function_exit?

Reply:
You mean have the scope end before record_function_exit runs? It's pretty hard to do that with the current thread_local_func & parent_ stack.